### DISCUSSION TITLE
ZFS pools cannot import after shrink the size

### DISCUSSION THREAD

**Discussion by xunge** [openzfs/zfs#13852] [Q&A]
I previously used the following command to expand the zfs pool:
```
sudo truncate -s +512G /var/snap/lxd/common/lxd/disks/lxd.img
sudo zpool set autoexpand=on lxd
sudo zpool online -e lxd /var/snap/lxd/common/lxd/disks/lxd.img
sudo zpool set autoexpand=off lxd
```
Since there was not enough disk space, I again used the following command to shrink the zfs pool:
```
sudo truncate -s -1G /var/snap/lxd/common/lxd/disks/lxd.img
sudo zpool set autoexpand=on lxd
sudo zpool online -e lxd /var/snap/lxd/common/lxd/disks/lxd.img
sudo zpool set autoexpand=off lxd
```
The zfs pool cannot be imported after reboot.
```
sudo zpool import -f -d /var/snap/lxd/common/lxd/disks lxd
cannot import 'lxd': I/O error
	Destroy and re-create the pool from
	a backup source.
```
[discussions/13745](https://github.com/openzfs/zfs/discussions/13745) indicates that zfs can't be reduced in size.

But the data in pool is very important.

I think there may be a problem with the file header that causes inconsistent data checks after reducing the zfs pool size.
Is it possible to recover pool by modifying the file header?

Or is there any way to extract the files from the zfs pool?

Thanks a lot!

**Comment by IvanVolosyuk** [comment]
> sudo truncate -s -1G /var/snap/lxd/common/lxd/disks/lxd.img

This is a destructive operation which destroys part of the pool data and corrupts it. Also, ZFS doesn't support pool shrinking.
Even with the filesystems which support shrinking you supposed to do that in the opposite way - shrink filesystem, shrink the device.

**Reply by xunge** [reply]
Thank you for reply, so is there a possibility to recover data?

**Reply by ryao** [reply]
It would require extracting the original file from the filesystem that stored ZFS. How to do this depends on the filesystem and is not always possible. You would want to ask the developers of the filesystem you use under ZFS.

So much time has elapsed that I would not think there is much hope.
